File: mobile/android/app/src/main/java/com/game_engine/casino/ui/tournament/TournamentViewModel.kt
package com.game_engine.casino.ui.tournament

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.UUID
import javax.inject.Inject

/**
 * Tournament ViewModel
 * 
 * Manages tournament browsing, registration, and in-tournament state.
 */
@HiltViewModel
class TournamentViewModel @Inject constructor(
    private val tournamentRepository: TournamentRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow<TournamentUiState>(TournamentUiState.Loading)
    val uiState: StateFlow<TournamentUiState> = _uiState.asStateFlow()

    private val _selectedTournament = MutableStateFlow<Tournament?>(null)
    val selectedTournament: StateFlow<Tournament?> = _selectedTournament.asStateFlow()

    private val _leaderboard = MutableStateFlow<List<LeaderboardEntry>>(emptyList())
    val leaderboard: StateFlow<List<LeaderboardEntry>> = _leaderboard.asStateFlow()

    init {
        loadTournaments()
    }

    fun loadTournaments() {
        viewModelScope.launch {
            _uiState.value = TournamentUiState.Loading
            try {
                val tournaments = tournamentRepository.getTournaments()
                _uiState.value = TournamentUiState.Success(tournaments)
            } catch (e: Exception) {
                _uiState.value = TournamentUiState.Error(e.message ?: "Failed to load tournaments")
            }
        }
    }

    fun selectTournament(tournamentId: String) {
        viewModelScope.launch {
            try {
                val tournament = tournamentRepository.getTournament(tournamentId)
                _selectedTournament.value = tournament
                
                // Load leaderboard
                val board = tournamentRepository.getLeaderboard(tournamentId)
                _leaderboard.value = board
            } catch (e: Exception) {
                // Handle error
            }
        }
    }

    fun registerForTournament(tournamentId: String) {
        viewModelScope.launch {
            try {
                tournamentRepository.register(tournamentId)
                // Refresh tournament details
                selectTournament(tournamentId)
            } catch (e: Exception) {
                // Handle error
            }
        }
    }

    fun unregisterFromTournament(tournamentId: String) {
        viewModelScope.launch {
            try {
                tournamentRepository.unregister(tournamentId)
                selectTournament(tournamentId)
            } catch (e: Exception) {
                // Handle error
            }
        }
    }

    fun filterTournaments(filter: TournamentFilter) {
        viewModelScope.launch {
            val tournaments = tournamentRepository.getTournaments(filter)
            _uiState.value = TournamentUiState.Success(tournaments)
        }
    }
}

sealed class TournamentUiState {
    object Loading : TournamentUiState()
    data class Success(val tournaments: List<Tournament>) : TournamentUiState()
    data class Error(val message: String) : TournamentUiState()
}

data class Tournament(
    val id: String,
    val name: String,
    val description: String,
    val gameType: String,
    val status: TournamentStatus,
    val startTime: LocalDateTime,
    val endTime: LocalDateTime,
    val buyIn: BigDecimal,
    val prizePool: BigDecimal,
    val maxPlayers: Int,
    val registeredPlayers: Int,
    val blindStructure: String,
    val isRegistered: Boolean = false,
    val currentLevel: Int = 1,
    val playerCount: Int = 0
)

enum class TournamentStatus {
    UPCOMING,
    REGISTRATION_OPEN,
    REGISTRATION_CLOSED,
    IN_PROGRESS,
    COMPLETED,
    CANCELLED
}

data class LeaderboardEntry(
    val rank: Int,
    val playerId: String,
    val playerName: String,
    val avatarUrl: String?,
    val chipCount: BigDecimal,
    val isCurrentPlayer: Boolean = false
)

enum class TournamentFilter {
    ALL,
    UPCOMING,
    ACTIVE,
    COMPLETED,
    MY_TOURNAMENTS
}

/**
 * Repository for tournament data
 */
class TournamentRepository {
    suspend fun getTournaments(filter: TournamentFilter = TournamentFilter.ALL): List<Tournament> {
        // Would call API
        return emptyList()
    }

    suspend fun getTournament(tournamentId: String): Tournament? {
        return null
    }

    suspend fun getLeaderboard(tournamentId: String): List<LeaderboardEntry> {
        return emptyList()
    }

    suspend fun register(tournamentId: String) {
        // API call
    }

    suspend fun unregister(tournamentId: String) {
        // API call
    }
}

File: mobile/ios/CasinoGame/Features/Tournament/TournamentViewModel.swift
import Foundation
import Combine

/**
 * Tournament View Model
 * 
 * Manages tournament state and operations.
 */
class TournamentViewModel: ObservableObject {
    
    @Published var uiState: TournamentUiState = .loading
    @Published var selectedTournament: Tournament?
    @Published var leaderboard: [LeaderboardEntry] = []
    
    private var cancellables = Set<AnyCancellable>()
    
    func loadTournaments() {
        uiState = .loading
        
        // API call would go here
        // For demo, simulate delay and empty list
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
            self?.uiState = .success([])
        }
    }
    
    func filterTournaments(_ filter: TournamentFilter) {
        // Filter and reload
    }
    
    func selectTournament(_ tournamentId: String) {
        // Load tournament details
    }
    
    func registerForTournament(_ tournamentId: String) {
        // API call to register
    }
    
    func unregisterFromTournament(_ tournamentId: String) {
        // API call to unregister
    }
}

// MARK: - Models

enum TournamentUiState {
    case loading
    case success([Tournament])
    case error(String)
}

struct Tournament: Identifiable {
    let id: String
    let name: String
    let gameType: String
    let status: TournamentStatus
    let startTime: Date
    let endTime: Date
    let buyIn: Double
    let prizePool: Double
    let maxPlayers: Int
    let registeredPlayers: Int
    var isRegistered: Bool = false
}

enum TournamentStatus: String {
    case upcoming
    case registrationOpen = "registration_open"
    case registrationClosed = "registration_closed"
    case inProgress = "in_progress"
    case completed
    case cancelled
}

struct LeaderboardEntry: Identifiable {
    let id = UUID()
    let rank: Int
    let playerName: String
    let avatarUrl: String?
    let chipCount: Double
    var isCurrentPlayer: Bool = false
}

enum TournamentFilter: Int, CaseIterable {
    case all
    case upcoming
    case active
    case myTournaments
}
